test/functional/legacy/normal_spec.lua

```diff
@@ -29,6 +35,71 @@ describe('normal', function()
       {8:    }44444444444444444444                |
                                               |
     ]])
-    end
-  )
+  end)
+
+  -- oldtest: Test_single_line_scroll()
+  it('(Half)-page scroll up or down reveals virtual lines #19605, #27967', function()
+    fn.setline(1, 'foobar one two three')
+    exec('set smoothscroll')
+    local ns = api.nvim_create_namespace('')
+    api.nvim_buf_set_extmark(0, ns, 0, 0, {
+      virt_lines = { { { '---', 'IncSearch' } } },
+      virt_lines_above = true,
+    })
+    -- Nvim: not actually necessary to scroll down to hide the virtual line.
+    -- Check topfill instead of skipcol and show the screen state.
+    feed('<C-E>')
+    eq(0, fn.winsaveview().topfill)
+    local s1 = [[
+      ^foobar one two three                    |
+      {1:~                                       }|*17
+                                              |
+    ]]
+    screen:expect(s1)
+    feed('<C-B>')
+    eq(1, fn.winsaveview().topfill)
+    local s2 = [[
+      {2:---}                                     |
+      ^foobar one two three                    |
+      {1:~                                       }|*16
+                                              |
+    ]]
+    screen:expect(s2)
+    feed('<C-E>')
+    eq(0, fn.winsaveview().topfill)
+    screen:expect(s1)
+    feed('<C-U>')
+    eq(1, fn.winsaveview().topfill)
+    screen:expect(s2)
+
+    -- Nvim: also test virt_lines below the last line
+    feed('yy100pG<C-L>')
+    api.nvim_buf_set_extmark(0, ns, 100, 0, { virt_lines = { { { '---', 'IncSearch' } } } })
+    screen:expect({
+      grid = [[
+        foobar one two three                    |*17
+        ^foobar one two three                    |
+                                                |
+      ]],
+    })
+    feed('<C-F>')
```

Comment:
Scrolling beyond the end of the buffer is different from the old behavior and slightly unexpected for (half)-page scrolling I think (more so than for CTRL-E/Y). I'll check if we can maintain the old behavior while scrolling down to reveal virtual lines.